File: python/21_merge_two_sorted_lists/solution_21_merge_two_sorted_lists.py
# Definition for singly-linked list.


class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution(object):
    def mergeTwoLists(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        if list1 == None and list2:
            return list2
        if list2 == None and list1:
            return list1
        result = ListNode(0)
        curr = result
        while list1 != None and list2 != None:
            if list1.val < list2.val:
                curr.next = list1
                list1 = list1.next
                curr = curr.next
                if list1 == None and list2:
                    curr.next = list2
                    break

            else:
                curr.next = list2
                list2 = list2.next
                curr = curr.next
                if list2 == None and list1:
                    curr.next = list1
                    break
        return result.next

File: typescript/21_Merge_Two_Sorted_Lists/solution_21_Merge_Two_Sorted_Lists.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
    const result = new ListNode(0);
    let currNode = result;
    while(list1 || list2) {
        if(list1 && list2) {
            if(list1.val <= list2.val) {
                currNode.next = list1
                list1 = list1.next
            } else {
                currNode.next = list2
                list2 = list2.next
            }
        } else if (!list1 && list2) {
            currNode.next = list2
            list2 = list2.next
        } else if (!list2 && list1) {
            currNode.next = list1
            list1 = list1.next
        }
        currNode = currNode.next
    }
    return result.next;
};